File: src/Java/algorithms/linkedlist/RemoveNthNodeFromEndOfList.java
package algorithms.linkedlist; 

import java.util.*; 
import lib.ListNode; 

public class RemoveNthNodeFromEndOfList {
    /* Function: Remove the nth node from the end of the list in one pass, where n is non-zero based. 
     * Solution: Be careful if n is larger than the list size, or the head is deleted. 
     */ 
    
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if (head == null || n <= 0) {
            return head; 
        }
        
        ListNode dummy = new ListNode(0); 
        dummy.next = head; 
        
        ListNode fast = dummy, slow = dummy; 
        for (int i = 0; i < n; ++i) {
            fast = fast.next; 
            if (fast == null) {
                return head; // So head would never be null
            }
        }
        
        while (fast.next != null) {
            fast = fast.next; 
            slow = slow.next; 
        }
        
        slow.next = slow.next.next; 
        return dummy.next; 
    }
}

File: src/CPP/src/leetcode/RemoveNthNodeFromEndOfList.cpp
/*
 * [Source] https://leetcode.com/problems/remove-nth-node-from-end-of-list/
 * [Difficulty]: Easy
 * [Tag]: Linked List
 * [Tag]: Two Pointers
 */

#include <iostream>

using namespace std;

// [Solution]: Use fast-slow pointers to find the previous node of the node that needs to be removed.
// [Corner Case]: If the node is the first node. So use dummy header.
class Solution {
};

/* Java solution
https://github.com/ZhenyingZhu/ClassicAlgorithms/blob/master/src/algorithms/linkedlist/RemoveNthNodeFromEndOfList.java
 */

/* Java solution
public class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if (head == null || n <= 0) {
            return head; 
        }
        ListNode fast = head; 
        while (n >= 0) {
            if (fast == null) {
                if (n == 0) {
                    return head.next; 
                } else {
                    return head; 
                }
            }
            fast = fast.next; 
            --n; 
        }
        ListNode cur = head; 
        while (fast != null) {
            cur = cur.next; 
            fast = fast.next; 
        }
        cur.next = cur.next.next; 
        return head; 
    }
}
 */


int main() {
    Solution sol;

    return 0;
}
